This is a refactor of the AWS infrastructure, including:

mikkoc opened the pull request:
- Create a public Load Balancer for the API server
- Create an internal Load Balancer for the API server
- Add a second public subnet, in a different AZ
- Assign a public IP to the Gandalf server
- Create DNS records for all the components
- Use the existing Hosted Zone instead of creating a new one every time

annashipman commented:
This PR is quite difficult to review because all the changes are in one large commit. Let's look at this together later today.

Commit:
Split out Gandalf server

We need to be able to reach the Gandalf server directly to push commits through SSH.
We tried to have the Gandalf instance behind an ELB but that did not work because it's storing persistent data on the instance so we need access to that specific instance, not wherever the ELB wants to send us.
Additionally, we cannot have a listener on port 22 on the ELB.

In this commit we:
- remove the Gandalf instance from the ones created via the loop
- create a new instance with a public IP
- create the DNS A record to point to the IP
- add a security group to allow connection on port 22

Commit:
Add external API load balancer

In order to distribute the load and have an HA resilent solution, we are adding an ELB, which will allow us to add more instances later on.

Commit:
Add internal API load balancer

In order to distribute the load and have an HA resilent solution, we are adding an internal ELB.

Commit:
Add second public subnet

In order to achieve HA and resiliency we are adding a second public subnet in a different Availability Zone.
This will allow us to use different subnets for the Tsuru components that support some sort of HA.

Commit:
USe a pre-existing Hosted Zone

Istead of creating a new hosted zone every time with Terraform, we're hardcoding a pre-exisisting Hosted Zone ID.
This has been already delegated in Dyn so it's actually the only working Hosted Zone.
Any new Hosted Zone created with Terraform would require some extra delegation steps in Dyn.

Commit:
Add CNAME wildcard DNS record

This record allows us to reach any application deployed to Tsuru without having to create a record for each application.

Commit:
Add a few DNS records for ease of use

Commit:
Use public subnets for the ELB

The ELB must reside in a public subnet in order for it to be reachable from the outside.

Commit:
Add ASG tags

At the moment instances created via Autoscaling Group do not get tagged. See: https://github.com/hashicorp/terraform/pull/1319
This is not implemented in Terraform 0.3.7 but as soon as we switch to 0.4.0 this will be available.

mikkoc commented:
We've split out the single commits, please have a look.

dcarley commented on app-servers.tf:
```diff
@@ -1,17 +1,60 @@
+}
+
+/* API internal Load balancer */
+resource "aws_elb" "api-int" {
```
What's the reason for having two separate ELBs? Could the same be achieved with one?

dcarley replied:
Nassim has helped answer this. It was created because it was thought that some internal components needed to talk to the API and they couldn't/shouldn't (?) go through the public facing ELB.

As I understand, by using the hostname's that AWS provides, split-horizon DNS will resolve to the appropriate external or internal address. Which means that security groups work as expected and you don't get charge for the traffic.

Nevertheless, it was later determined that the internal components don't call the API and that we can refactor these two ELBs out. That should be done before merging this PR.

dcarley replied:
> As I understand, by using the hostname's that AWS provides, split-horizon DNS will resolve to the appropriate external or internal address. Which means that security groups work as expected and you don't get charge for the traffic.

I understood/remembered wrong. ELB hostnames don't seem to use split-horizon like instances do.

mikkoc replied:
As far as I know, the internal API is being used in Gandalf.
In particular, in the ansible role, the task: `Add api endpoint to bash_profile`

dcarley replied:
:+1:

dcarley commented on security-groups.tf:
```diff
@@ -35,6 +35,24 @@ resource "aws_security_group" "nat" {
+    from_port = 22
+    to_port   = 22
+    protocol  = "tcp"
+    cidr_blocks = ["80.194.77.0/24"]
```
This is too wide - we don't own all of this range.

dcarley replied:
Luigi has noted that this is consistent with the other entries in this file. So we're going to merge as-is and tackle them in a separate PR.

annashipman replied:
@dcarley @lucorsel Do we have a PR to tackle this already? If not, do we have a story to address it?

mikkoc replied:
Is there an official list of GDS IPs we should be using?
If yes, I can go ahead and create the PR. I remember we were using /32 in the beginning but we were getting occasional issues connecting. I believe even if we're all using the same `CDN02_test` wi-fi network there's a handful of /32 IPs that come up.
